google/cloud/pubsub/message.h

```diff
@@ -0,0 +1,194 @@
+ public:
+  /// Change the payload in @p m to @p data
+  static Message SetData(Message m, std::string data) {
+    m.proto_.set_data(std::move(data));
+    return m;
+  }
+
+  /// Change the attributes in @p m to those in the range [@p begin, @p end)
+  template <typename Iterator>
+  static Message SetAttributes(Message m, Iterator begin, Iterator end) {
+    using value_type =
+        google::protobuf::Map<std::string, std::string>::value_type;
+    m.proto_.clear_attributes();
+    for (auto kv = begin; kv != end; ++kv) {
+      using std::get;
+      m.proto_.mutable_attributes()->insert(
+          value_type(get<0>(*kv), get<1>(*kv)));
+    }
+    return m;
+  }
+
+  /// Change the attributes in @p m to the values in @p v
+  static Message SetAttributes(
+      // NOLINTNEXTLINE(performance-unnecessary-value-param)
+      Message m, std::vector<std::pair<std::string, std::string>> v) {
+    using value_type =
+        google::protobuf::Map<std::string, std::string>::value_type;
+    m.proto_.clear_attributes();
+    for (auto& kv : v) {
+      m.proto_.mutable_attributes()->insert(
+          value_type(std::move(kv.first), std::move(kv.second)));
+    }
+    return m;
+  }
+
+  /// Change the attributes in @p m to the values in @p v
+  template <typename Pair>
+  static Message SetAttributes(Message m, std::vector<Pair> v) {
+    return SetAttributes(std::move(m), v.begin(), v.end());
+  }
+
+  /// Create a new message with @p data as its payload
+  static Message FromData(std::string data) {
+    return SetData(Message{}, std::move(data));
+  }
+
+  /// Create a new message with the attributes in the range [@p begin, @p end)
+  template <typename Iterator>
+  static Message FromAttributes(Iterator begin, Iterator end) {
+    return SetAttributes(Message{}, begin, end);
+  }
+
+  /// Create a new message with the attributes in @p v
+  static Message FromAttributes(
+      std::vector<std::pair<std::string, std::string>> v) {
+    return SetAttributes(Message{}, std::move(v));
+  }
+
+  /// Create a new message with the attributes in @p v
+  template <typename Pair>
+  static Message FromAttributes(std::vector<Pair> v) {
+    return SetAttributes(Message{}, std::move(v));
+  }
+
+  /// @name accessors
+  //@{
+  std::string const& data() const& { return proto_.data(); }
+  std::string&& data() && { return std::move(*proto_.mutable_data()); }
+  std::string const& message_id() const { return proto_.message_id(); }
+  std::string const& ordering_key() const { return proto_.ordering_key(); }
+  std::chrono::system_clock::time_point publish_time() const;
+  std::map<std::string, std::string> attributes() const {
+    std::map<std::string, std::string> r;
+    for (auto const& kv : proto_.attributes()) {
+      r.emplace(kv.first, kv.second);
+    }
+    return r;
+  }
+  //@}
+
+  /// @name Copy and move
+  //@{
+  Message(Message const&) = default;
+  Message& operator=(Message const&) = default;
+  Message(Message&&) = default;
```

Comment:
Well, quasi-immutable, the only mutation was "move".  I mean, sure, you can do a move and continue to use the object, but that is just bad manners.
Anyway, no longer relevant, since I moved this to use a builder as you suggested.